Increase test coverage for PG15 features

hanefi opened the pull request:
This PR aims to add more tests for covering various features of PG15.

Instead of creating a large PR, I prefer to merge smaller PRs.

onderkalaci commented on src/test/regress/expected/pg15.out:
```diff
@@ -388,15 +388,113 @@ ON (true)
+(1 row)
+
+INSERT into numeric_test SELECT x,x FROM generate_series(111, 115) x;
+SELECT * FROM numeric_test ORDER BY 1,2;
```
I think we do have a critical issue with shard pruning on this:
```SQL



CREATE TABLE numeric_test(id numeric(3,-1), orig_value int);
SELECT create_distributed_table('numeric_test','id');
INSERT into numeric_test SELECT x,x FROM generate_series(111, 115) x;

SELECT * FROM numeric_test;
┌─────┬────────────┐
│ id  │ orig_value │
├─────┼────────────┤
│ 120 │        115 │
│ 110 │        112 │
│ 110 │        113 │
│ 110 │        114 │
│ 110 │        111 │
└─────┴────────────┘
(5 rows)

-- OPPPPS where is the row?
 SELECT * FROM numeric_test  WHERE id  = 120;
┌────┬────────────┐
│ id │ orig_value │
├────┼────────────┤
└────┴────────────┘
(0 rows)

SELECT undistribute_table('numeric_test');
NOTICE:  creating a new table for public.numeric_test
NOTICE:  moving the data of public.numeric_test
NOTICE:  dropping the old public.numeric_test
NOTICE:  renaming the new table to public.numeric_test
┌────────────────────┐
│ undistribute_table │
├────────────────────┤
│                    │
└────────────────────┘
(1 row)


 SELECT * FROM numeric_test  WHERE id  = 120;
┌─────┬────────────┐
│ id  │ orig_value │
├─────┼────────────┤
│ 120 │        115 │
└─────┴────────────┘
(1 row)

```

For example, if you do `INSERT into numeric_test SELECT x,x FROM generate_series(111, 115) x;` before create-dist-table, this works.

Needs more investigation.

Could be related to https://github.com/citusdata/citus/pull/2640 or needs something more for this type

hanefi replied:
I tried to reproduce a similar issue with earlier versions of PG today. I was not able to do that. My guess is that we need negative scales to have this problem.

onderkalaci replied:
If the fix is non-trivial, it is acceptable to throw an error if dist key is this non-supported numeric type. We can consider prioritizing that if/when needed

onderkalaci replied:
Also, what about repartitioning on such columns? If that also doesn't work, I'm fine with not allowing create dist / ref table if the table include these columns

naisila reviewed (changes requested):
These tests look good to me. 🚀 

Let's add some tests about numeric types with scale greater than the precision as well, then we can merge.

hanefi commented:
> Let's add some tests about numeric types with scale greater than the precision as well, then we can merge.

I added some tests for these. I did some more changes that may not be obvious, so let me list them here.

- I dropped the commits for SQL/JSON as those commits are now reverted on PG15
- I dropped a commit for numeric types as I merged that as part of another PR
- I added a test for an expected `alter_distributed_table` failure on a numeric column with a negative scale.
- I needed to add a normalization rule for `alter_distributed_table` as the generated table names are known to change.
